crates/iota-data-ingestion-core/src/worker_pool.rs

```diff
@@ -92,42 +110,88 @@ impl<W: Worker + 'static> WorkerPool<W> {
+                                        }
+                                }
                             }
-                            current_checkpoint_number += 1;
-                        }
-                        if let Some(update) = executor_status_update {
-                            if executor_progress_sender
-                                .send((self.task_name.clone(), update))
-                                .await.is_err() {
-                                    // The executor progress channel closing is a sign we need to
-                                    // exit this loop.
+                            // By checking if token was not cancelled we ensure that no
+                            // further checkpoints will be sent to the workers
+                            while !token.is_cancelled() && !checkpoints.is_empty() && !idle.is_empty() {
+                                let checkpoint = checkpoints.pop_front().unwrap();
+                                let worker_id = idle.pop_first().unwrap();
+                                if workers[worker_id].send(checkpoint).await.is_err() {
+                                    // The worker channel closing is a sign we need to exit this loop.
                                     break;
                                 }
+                            }
                         }
-                    }
-                    while !checkpoints.is_empty() && !idle.is_empty() {
-                        let checkpoint = checkpoints.pop_front().unwrap();
-                        let worker_id = idle.pop_first().unwrap();
-                        if workers[worker_id].0.send(checkpoint).await.is_err() {
-                            // The worker channel closing is a sign we need to exit this loop.
-                            break;
+                        // Manages the graceful shutdown sequence of the worker pool.
+                        //
+                        // The shutdown process follows these steps:
+                        // 1. Token cancellation triggers workers to:
+                        //    - Complete their current checkpoint processing
+                        //    - Send a final ShutDown message
+                        //    - Exit their processing loop
+                        //
+                        // 2. Main loop continues running to:
+                        //    - Process any remaining Progress messages
+                        //    - Track worker shutdowns via ShutDown messages
+                        //    - Once all workers have reported shutdown, await their join handles
+                        //
+                        // This ensures:
+                        // - No work is interrupted mid-processing
+                        // - All progress messages are processed
+                        // - All resources are properly cleaned up
+                        // - Workers terminate in an orderly fashion
```

Comment:
I would suggest transferring to the module-level documentation.

Reply:
Since the README.md has a section on graceful shutdown could that be enough?

Reply:
[fixed](https://github.com/iotaledger/iota/pull/4648/commits/af78d4b08780698c81f8610bb0683cee2d17ec59)